projects/rocr-runtime/rocrtst/suites/functional/memory_accounting.cc

```diff
@@ -0,0 +1,211 @@
+    if (access == HSA_AMD_MEMORY_POOL_ACCESS_NEVER_ALLOWED) return;
+
+    rocrtst::pool_info_t pool_info;
+    err = rocrtst::AcquirePoolInfo(pool, &pool_info);
+    if (err != HSA_STATUS_SUCCESS || !pool_info.alloc_allowed) return;
+
+    if (pool_info.segment != HSA_AMD_SEGMENT_GLOBAL) return;
+
+    hsa_device_type_t device_type;
+    err = hsa_agent_get_info(agent, HSA_AGENT_INFO_DEVICE, &device_type);
+    if (err != HSA_STATUS_SUCCESS || device_type != HSA_DEVICE_TYPE_GPU) return;
+
+    bool vmem_supported = false;
+    err = hsa_system_get_info(HSA_AMD_SYSTEM_INFO_VIRTUAL_MEM_API_SUPPORTED, &vmem_supported);
+    if (err != HSA_STATUS_SUCCESS || !vmem_supported) return;
+
+    rocrtst::pool_info_t pool_i;
+
+    ASSERT_SUCCESS(rocrtst::AcquirePoolInfo(pool, &pool_i));
+
+    size_t granule_size = pool_i.alloc_rec_granule;
+    size_t allocation_size = 10 * granule_size;
+
+    size_t amount_begin = 0, amount_current = 0;
+    void* reserved_addr = nullptr;
+    hsa_amd_vmem_alloc_handle_t mem_handle;
+
+    ASSERT_SUCCESS(hsa_agent_get_info(agent, (hsa_agent_info_t)HSA_AMD_AGENT_INFO_MEMORY_AVAIL, &amount_begin));
+    ASSERT_SUCCESS(hsa_amd_vmem_address_reserve(&reserved_addr, allocation_size, 0, 0));
+
+    hsa_amd_pointer_info_t ptr_info = {};
+    ptr_info.size = sizeof(ptr_info);
+    err = hsa_amd_pointer_info(reserved_addr, &ptr_info, nullptr, nullptr, nullptr);
+    if (err != HSA_STATUS_SUCCESS || ptr_info.type != HSA_EXT_POINTER_TYPE_RESERVED_ADDR) {
+        hsa_amd_vmem_address_free(reserved_addr, allocation_size);
+        return;
+    }
+
+    err = hsa_amd_vmem_handle_create(pool, allocation_size, MEMORY_TYPE_NONE, 0, &mem_handle);
+    if (err != HSA_STATUS_SUCCESS) {
+        hsa_amd_vmem_address_free(reserved_addr, allocation_size);
+        return;
+    }
+
+    err = hsa_amd_vmem_map(reserved_addr, allocation_size, 0, mem_handle, 0);
+    if (err != HSA_STATUS_SUCCESS) {
+        hsa_amd_vmem_handle_release(mem_handle);
+        hsa_amd_vmem_address_free(reserved_addr, allocation_size);
+        return;
+    }
+
+    hsa_amd_memory_access_desc_t access_desc = {HSA_ACCESS_PERMISSION_RW, agent};
+    err = hsa_amd_vmem_set_access(reserved_addr, allocation_size, &access_desc, 1);
+    if (err != HSA_STATUS_SUCCESS) {
+        hsa_amd_vmem_unmap(reserved_addr, allocation_size);
+        hsa_amd_vmem_handle_release(mem_handle);
+        hsa_amd_vmem_address_free(reserved_addr, allocation_size);
+        return;
+    }
+
+    ASSERT_SUCCESS(hsa_agent_get_info(agent, (hsa_agent_info_t)HSA_AMD_AGENT_INFO_MEMORY_AVAIL, &amount_current));
+    ASSERT_NEAR(amount_begin - amount_current, allocation_size, 4096);
+
+    ASSERT_SUCCESS(hsa_amd_vmem_unmap(reserved_addr, allocation_size));
+    ASSERT_SUCCESS(hsa_amd_vmem_handle_release(mem_handle));
+
+    ASSERT_SUCCESS(hsa_agent_get_info(agent, (hsa_agent_info_t)HSA_AMD_AGENT_INFO_MEMORY_AVAIL, &amount_current));
+    ASSERT_EQ(amount_begin, amount_current);
```

Comment:
Please move the memory-avail check after the hsa_amd_vmem_address_free.

Ideally the memory-avail should be back to the original value after hsa_amd_vmem_handle_release, but it seems the extra CPU mapping is causing KFD to only free the memory after the VA is released. Not great, but that is ok for now.